openvm/extensions/hints-circuit/src/lib.rs

```diff
@@ -0,0 +1,248 @@
+use openvm_circuit::derive::{AnyEnum, InstructionExecutor};
+use openvm_circuit::system::memory::MemoryController;
+use openvm_circuit::system::phantom::PhantomChip;
+use openvm_instructions::riscv::RV32_MEMORY_AS;
+use openvm_instructions::PhantomDiscriminant;
+use openvm_rv32im_circuit::adapters::unsafe_read_rv32_register;
+use openvm_stark_backend::p3_field::PrimeField32;
+use powdr_openvm_hints_transpiler::HintsPhantom;
+
+// this module is mostly copy/pasted code from k256 for the field element representation in 32-bit architectures
+mod field_k256;
+
+pub struct HintsExtension;
+
+#[derive(ChipUsageGetter, Chip, InstructionExecutor, AnyEnum)]
+pub enum HintsExecutor<F: PrimeField32> {
+    // TODO: not sure what to do when there's no "real" instruction in the extension
+    Phantom(PhantomChip<F>),
+}
+
+#[derive(ChipUsageGetter, Chip, AnyEnum)]
+pub enum HintsPeriphery<F: PrimeField32> {
+    Phantom(PhantomChip<F>),
+}
+
+impl<F: PrimeField32> VmExtension<F> for HintsExtension {
+    type Executor = HintsExecutor<F>;
+    type Periphery = HintsPeriphery<F>;
+
+    fn build(
+        &self,
+        builder: &mut openvm_circuit::arch::VmInventoryBuilder<F>,
+    ) -> Result<
+        openvm_circuit::arch::VmInventory<Self::Executor, Self::Periphery>,
+        openvm_circuit::arch::VmInventoryError,
+    > {
+        let inventory = VmInventory::new();
+        builder.add_phantom_sub_executor(
+            ReverseBytesSubEx,
+            PhantomDiscriminant(HintsPhantom::HintReverseBytes as u16),
+        )?;
+        builder.add_phantom_sub_executor(
+            K256InverseFieldSubEx,
+            PhantomDiscriminant(HintsPhantom::HintK256InverseField as u16),
+        )?;
+        builder.add_phantom_sub_executor(
+            K256InverseField10x26SubEx,
+            PhantomDiscriminant(HintsPhantom::HintK256InverseField10x26 as u16),
+        )?;
+        builder.add_phantom_sub_executor(
+            K256SqrtField10x26SubEx,
+            PhantomDiscriminant(HintsPhantom::HintK256SqrtField10x26 as u16),
+        )?;
+        Ok(inventory)
+    }
+}
+
+pub struct ReverseBytesSubEx;
+
+impl<F: PrimeField32> PhantomSubExecutor<F> for ReverseBytesSubEx {
+    fn phantom_execute(
+        &mut self,
+        memory: &MemoryController<F>,
+        streams: &mut Streams<F>,
+        _discriminant: PhantomDiscriminant,
+        a: F,
+        _b: F,
+        c_upper: u16,
+    ) -> eyre::Result<()> {
+        assert_eq!(c_upper, 0);
+        // read register
+        let rs1 = unsafe_read_rv32_register(memory, a);
+        // read memory
+        let bytes = memory.unsafe_read::<4>(
+            F::from_canonical_u32(RV32_MEMORY_AS),
+            F::from_canonical_u32(rs1),
+        );
+        // write hint as bytes in reverse
+        let hint_bytes = bytes.into_iter().rev().collect();
+        // let hint_bytes = std::iter::once(F::from_canonical_u32(42)).collect();
+        streams.hint_stream = hint_bytes;
```

Comment:
Just guessing here, but are we sure that we don't need the previous value of `streams.hint_stream`? Or is it to allow passing data from different phantom sub executors? Can we assert the previous value to be something?

Reply:
I think OpenVM does the same thing: every time `hint_stream` is needed it is emptied and copied from `input_stream`, which actually contains all values passed via `StdIn`.